src/accelerate/accelerator.py

```diff
@@ -367,6 +380,39 @@ def no_sync(self, model):
         with context():
             yield
 
+    @property
+    def _do_sync(self) -> bool:
+        "Checks if gradients should be synchronized and the optimizers + schedulers should be stepped"
+        if self.state.end_of_dataloader:
```

Comment:
This shouldn't be in the state, the state only should contain the environment info, not some things that are subject to change. So store it directly in self.